### ISSUE TITLE
Issues about the implementation of the **soft** cross entropy loss for LwF

### ISSUE THREAD

**Issue by haizhou-shi** [aimagelab/mammoth#22]
Hi, amazing work! 

I have a question about the implementation of the cross entropy loss for the soft labels. 

In the LwF, you used the following code:
```
def smooth(logits, temp, dim):
    log = logits ** (1 / temp)
    return log / torch.sum(log, dim).unsqueeze(1)

def modified_kl_div(old, new):
    return -torch.mean(torch.sum(old * torch.log(new), 1))
```

and when you compute the loss between two logits, you use
```
modified_kl_div(smooth(self.soft(logits_a), 2, 1),
                             smooth(self.soft(logits_b), 2, 1))
```

I was wondering if there is a particular reason you did this instead of 
```
torch.mean(softmax(logits_a/temp) * logsoftmax(logits/temp))
```

Thank you so much and looking forward to your response!

**Comment by mbosc** [comment]
Hi there and thank you for the kind words.
Our current version of the code is a straight up transcription of the formulae on the [LwF paper](https://arxiv.org/pdf/1606.09282.pdf). In particular, the temperature scaling implemented by `smooth` can be found on Eq.4 and `modfied_kl_div` translates Eq.3.
Your suggested implementation does not look like it implements what we see in the paper, the temperature for sure is scaled exponentially and not by division as you suggest.

**Comment by haizhou-shi** [comment]
Hi, thank you for your timely response! 

Actually I did a little derivation myself and I found out the two implementations are mathematically equivalent. However, the original version of KD (proposed by Hinton et al., suggested as above), is more computationally stable (not directly dealing with log_softmax ourselves).

Maybe you can consider changing the LwF's implementation in the future? 

Best regards.

**Comment by mbosc** [comment]
Yes, we know that Hinton's formulation is preferable (and indeed we use that one in DER). We originally had an implementation of LwF similar to your suggestion, however we later opted for being maximally compatible with the original paper, so that we can use their hyperparameters when possible.

**Comment by haizhou-shi** [comment]
Cool! Understood. Thank you so much!

**Comment by loribonna** [comment]
Hi @MagicHealer , thanks for the question. I will look into it and let you know.

**Comment by loribonna** [comment]
@MagicHealer I did not find any issue reproducing the results of [der](https://arxiv.org/pdf/2004.07211.pdf). If you seek to reproduce the results of [x-der](https://arxiv.org/pdf/2201.00766.pdf) however you will need **LwF.MC**, which uses an implementation borrowed from the [icarl](https://arxiv.org/pdf/1611.07725.pdf) paper (it is basically iCaRL but with no buffer). Unfortunately, such method was not included in the current codebase but I will add it now (#24).

**Comment by haizhou-shi** [comment]
@MagicHealer I have no problem replicating the results of LwF in rot-MNIST and perm-MNIST, what is your result specifically?

**Comment by MagicHealer** [comment]
@haizhou-shi @loribonna I am trying to use Learning without forgetting using ImageNet as first task and CUB dataset as second task.

**Comment by loribonna** [comment]
@MagicHealer as commented in #23, this is not possible as of now in this version of Mammoth. To do that, you will need to implement a custom dataset following, for example, the specification in `datasets/seq_cifar10`.
